File: 0022-generate-parentheses/0022-generate-parentheses.java
class Solution {
    public List<String> generateParenthesis(int n) {
   List<String> ans=new ArrayList<>();
    solve(ans,n,0,0,new StringBuilder());
    return ans;
    }
    void solve(List<String> ans,int n,int o,int c,StringBuilder sb){
        if(sb.length()==n*2){
            ans.add(sb.toString());
            return;
        }
        if(o<n){
            sb.append('(');
            solve(ans,n,o+1,c,sb);
            sb.setLength(sb.length()-1);
        }
        if(c<o){
          sb.append(')');
            solve(ans,n,o,c+1,sb);
            sb.setLength(sb.length()-1);  
        }
    }
}

File: 0022-generate-parentheses/0022-generate-parentheses.py
class Solution:
    def __init__(self):
        self.l=[]
    def generateParenthesis(self, n: int) -> List[str]:
        self.solve(n,"",0,0)
        return self.l
    def solve(self,n,s,c,r):
        if (c+r==2*n):
            self.l.append(s)
        if c<n:     
            self.solve(n,s+'(',c+1,r)        
        if r<c:       
            self.solve(n,s+')',c,r+1)
          



        